giraffe/src/utils/getTicks.test.ts

```diff
@@ -1,13 +1,293 @@
+        ).toEqual(0)
+
+        tickStep = end - start + 1
+        expect(
+          generateTicks(valueDomain, VALUE, totalTicks, null, tickStep).length
+        ).toEqual(0)
+      })
+
+      it('should generate ticks accordingly when given total ticks, tick start, and tick step', () => {
+        const [start, end] = valueDomain
+        const [timeStart, timeEnd] = timeDomain
+        let totalTicks = 10
+        let tickStart = start
+        let tickStep = (end - start) / totalTicks
+        let result = generateTicks(
+          valueDomain,
+          VALUE,
+          totalTicks,
+          tickStart,
+          tickStep
+        )
+        expect(result.length).toBeGreaterThan(0)
+        expect(result.length).toBeLessThanOrEqual(totalTicks)
+
+        tickStart = timeStart
+        tickStep = (timeEnd - timeStart) / totalTicks
+        result = generateTicks(
+          timeDomain,
+          TIME,
+          totalTicks,
+          tickStart,
+          tickStep
+        )
+        expect(result.length).toBeGreaterThan(0)
+        expect(result.length).toBeLessThanOrEqual(totalTicks)
+
+        totalTicks = 10
+        tickStart = end
+        tickStep = ((end - start) / (totalTicks + 1)) * -1
+        result = generateTicks(
+          valueDomain,
+          VALUE,
+          totalTicks,
+          tickStart,
+          tickStep
+        )
+        expect(result.length).toBeGreaterThan(0)
+        expect(result.length).toBeLessThanOrEqual(totalTicks)
+
+        tickStart = timeEnd
+        tickStep = ((timeEnd - timeStart) / (totalTicks + 1)) * -1
+        result = generateTicks(
+          timeDomain,
+          TIME,
+          totalTicks,
+          tickStart,
+          tickStep
+        )
+        expect(result.length).toBeGreaterThan(0)
+        expect(result.length).toBeLessThanOrEqual(totalTicks)
+
+        expect(() => {
+          tickStep = 0
+          expect(
+            generateTicks(valueDomain, VALUE, totalTicks, tickStart, tickStep)
+              .length
+          ).toEqual(0)
+        }).not.toThrow()
+
+        totalTicks = 1
+        tickStart = start
+        tickStep = 0.01
+        result = generateTicks(
+          valueDomain,
+          VALUE,
+          totalTicks,
+          tickStart,
+          tickStep
+        )
+        expect(result.length).toEqual(totalTicks)
+
+        totalTicks = 1
+        tickStart = timeDomain[0]
+        tickStep = 1
+        result = generateTicks(
+          timeDomain,
+          TIME,
+          totalTicks,
+          tickStart,
+          tickStep
+        )
+        expect(result.length).toEqual(totalTicks)
+
+        totalTicks = 15
+        tickStart = start - 1
+        tickStep = end - start + 1
+        result = generateTicks(
+          valueDomain,
+          VALUE,
+          totalTicks,
+          tickStart,
+          tickStep
+        )
+        expect(result.length).toEqual(0)
+      })
+    })
+  })
```

Comment:
great tests from a tactical level. from a higher, strategic level, just reinforcing what i said before - there's a lot of interplay here that is pretty complex and will probably be really confusing for users and might be an indication of larger design issues.